Unable to login to application behind oauth-proxy when using public SSL certificate

I am unable to login to an application (prometheus) which is protected by oauth-proxy. The public URL of the application (and of the openshift console) is served by a SSL certificate signed by public authority (which is not the kubernetes CA).

When deploying the same application in toy environment (using the default router certificate, which is generated by the kubernetes CA) I have no problem logging in.

However, on the production system, I get an error 500 after submitting my credentials on the openshift logging page.

This is the logs of my prometheus pod (oauth-proxy container):
```sh
[root@customerdvrs01 ~]# oc logs -f prometheus-core-2-vql0t -c oauth-proxy 
2017/10/22 13:27:32 provider.go:476: Performing OAuth discovery against https://172.30.0.1/.well-known/oauth-authorization-server
2017/10/22 13:27:32 provider.go:522: 200 GET https://172.30.0.1/.well-known/oauth-authorization-server {
  "issuer": "https://openshift.customer.domain:8443",
  "authorization_endpoint": "https://openshift.customer.domain:8443/oauth/authorize",
  "token_endpoint": "https://openshift.customer.domain:8443/oauth/token",
  "scopes_supported": [
    "user:check-access",
    "user:full",
    "user:info",
    "user:list-projects",
    "user:list-scoped-projects"
  ],
  "response_types_supported": [
    "code",
    "token"
  ],
  "grant_types_supported": [
    "authorization_code",
    "implicit"
  ],
  "code_challenge_methods_supported": [
    "plain",
    "S256"
  ]
}
2017/10/22 13:27:32 provider.go:265: Delegation of authentication and authorization to OpenShift is enabled for bearer tokens and client certificates.
2017/10/22 13:27:32 oauthproxy.go:161: mapping path "/" => upstream "http://localhost:9090"
2017/10/22 13:27:32 oauthproxy.go:184: compiled skip-auth-regex => "^/metrics"
2017/10/22 13:27:32 oauthproxy.go:190: OAuthProxy configured for  Client ID: system:serviceaccount:monitoring:prometheus
2017/10/22 13:27:32 oauthproxy.go:200: Cookie settings: name:_oauth_proxy secure(https):true httponly:true expiry:168h0m0s domain:<default> refresh:disabled
2017/10/22 13:27:32 http.go:56: HTTP: listening on :8080
2017/10/22 13:27:50 oauthproxy.go:582: error redeeming code (client:10.128.0.1:44912): Post https://openshift.customer.domain:8443/oauth/token: x509: certificate signed by unknown authority
2017/10/22 13:27:50 oauthproxy.go:399: ErrorPage 500 Internal Error Internal Error
```

From the currently open PRs, I can see that there is an issue with not being able to use the system CA roots at the same time as the kubernetes CA; is my problem related and would merging PR #25 or #27 help?

--- a/http.go
+++ b/http.go
@@ -83,7 +83,7 @@ func (s *Server) ServeHTTPS() {
 
 	if len(s.Opts.TLSClientCAFiles) > 0 {
 		config.ClientAuth = tls.RequestClientCert
-		config.ClientCAs, err = util.GetCertPool(s.Opts.TLSClientCAFiles)
+		config.ClientCAs, err = util.GetCertPool(s.Opts.TLSClientCAFiles, false)
 		if err != nil {
 			log.Fatalf("FATAL: %s", err)
 		}

--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -104,7 +104,7 @@ func NewReverseProxy(target *url.URL, upstreamFlush time.Duration, rootCAs []str
 		IdleConnTimeout:     1 * time.Minute,
 	}
 	if len(rootCAs) > 0 {
-		pool, err := util.GetCertPool(rootCAs)
+		pool, err := util.GetCertPool(rootCAs, false)
 		if err != nil {
 			return nil, err
 		}

--- a/providers/openshift/provider.go
+++ b/providers/openshift/provider.go
@@ -76,13 +76,8 @@ func (p *OpenShiftProvider) LoadDefaults(serviceAccount string, caPaths []string
 	}
 	p.reviews = reviews
 
-	if len(caPaths) == 0 {
-		// ignore errors
-		p.SetCA([]string{"/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"})
-	} else {
-		if err := p.SetCA(caPaths); err != nil {
-			return nil, err
-		}
+	if err := p.setCA(caPaths); err != nil {
+		return nil, err
 	}
 
 	defaults := &providers.ProviderData{
@@ -119,14 +114,21 @@ func (p *OpenShiftProvider) LoadDefaults(serviceAccount string, caPaths []string
 }
 
 // SetCA initializes the client used for connecting to the master.
-func (p *OpenShiftProvider) SetCA(paths []string) error {
+func (p *OpenShiftProvider) setCA(paths []string) error {
 	if p.Client == nil {
 		p.Client = &http.Client{
 			Jar:       http.DefaultClient.Jar,
 			Transport: http.DefaultTransport,
 		}
 	}
-	pool, err := util.GetCertPool(paths)
+	//defaults
+	capaths := []string{"/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"}
+	system_roots := true
+	if len(paths) != 0 {
+		capaths = paths
+		system_roots = false
+	}
+	pool, err := util.GetCertPool(capaths, system_roots)
 	if err != nil {
 		return err
 	}

--- a/util/util.go
+++ b/util/util.go
@@ -4,13 +4,24 @@ import (
 	"crypto/x509"
 	"fmt"
 	"io/ioutil"
+	"log"
 )
 
-func GetCertPool(paths []string) (*x509.CertPool, error) {
+func GetCertPool(paths []string, system_roots bool) (*x509.CertPool, error) {
 	if len(paths) == 0 {
 		return nil, fmt.Errorf("Invalid empty list of Root CAs file paths")
 	}
-	pool := x509.NewCertPool()
+	var pool *x509.CertPool
+	if system_roots {
+		// ignore errors
+		pool, _ = x509.SystemCertPool()
+		if pool == nil {
+			log.Printf("No system certificates found")
+			pool = x509.NewCertPool()
+		}
+	} else {
+		pool = x509.NewCertPool()
+	}
 	for _, path := range paths {
 		data, err := ioutil.ReadFile(path)
 		if err != nil {

